torchvision/prototype/transforms/_augment.py

```diff
@@ -243,11 +239,12 @@ def _copy_paste(
         if blending:
             paste_alpha_mask = F.gaussian_blur(paste_alpha_mask.unsqueeze(0), kernel_size=[5, 5], sigma=[2.0])
 
+        inverse_paste_alpha_mask = ~paste_alpha_mask
```

Comment:
I just thought of another "evil hack" that avoids the out of place inversion here. There are two places where `paste_alpha_mask` or its inverse is used:

1. `image = (image * inverse_paste_alpha_mask).add_(paste_image * paste_alpha_mask)`
2. `masks = masks * inverse_paste_alpha_mask`

By reordering the first computation to

```py
image = (paste_image * paste_alpha_mask).add_(image * inverse_paste_alpha_mask)
```

we no longer need the `paste_alpha_mask` after `inverse_paste_alpha_mask` is used the first time. Thus, we can perform inplace operations on it:

```py
image = (paste_image * paste_alpha_mask).add_(image * paste_alpha_mask.logical_not_())
```

That turns the 2. statement into

```py
`masks = masks * paste_alpha_mask`
```

Although that looks wrong, it is actually correct since `paste_alpha_mask` at this point contains it inverse. 

Going for this needs at least an extensive comment since it is not obvious. Not sure if it is worth it.

Reply:
@pmeier I'm not 100% sure but I sense that this will end up having the same number of operations as before. You indeed do an inplace operation but at the end you still need to reverse the bool an write the results back to memory. I think what you propose will lead to less memory footprint. Still I think your proposal is worth trying. Do you want to bring it as an optimization on a PR after this one? You could refactor it a bit to split this into multiple lines and avoid doing one non-inplace op with the inplace op on `paste_alpha_mask` to make things clearer to the reader (that's debatable, maybe it's clear already?). You could literally measure only those 2 lines on the new PR and show-case if the proposal is faster. WDYT?

Reply:
You are right:

```py
def current(*, image, masks, paste_image, paste_alpha_mask, size1_ne_size2):
    inverse_paste_alpha_mask = paste_alpha_mask.logical_not()

    # Copy-paste images:
    image = image.mul(inverse_paste_alpha_mask).add_(paste_image.mul(paste_alpha_mask))
    # Copy-paste masks:
    masks = masks * inverse_paste_alpha_mask


def proposal(*, image, masks, paste_image, paste_alpha_mask, size1_ne_size2):
    if size1_ne_size2:
        tmp = paste_image.mul_(paste_alpha_mask)
    else:
        tmp = paste_image.mul(paste_alpha_mask)
    inverse_paste_alpha_mask = paste_alpha_mask.logical_not_()

    # Copy-paste images:
    image = tmp.add_(image.mul(inverse_paste_alpha_mask))
    # Copy-paste masks:
    masks = masks.mul(inverse_paste_alpha_mask)
```

<details><summary>Benchmark script</summary>
<p>

```py
import itertools

import torch
from torch.utils import benchmark

shapes = [
    (3, 512, 512),  # single image
    (5, 3, 512, 512),  # single video
]
dtypes = [torch.uint8, torch.float32]
devices = ["cpu", "cuda"]


def current(*, image, masks, paste_image, paste_alpha_mask, size1_ne_size2):
    inverse_paste_alpha_mask = paste_alpha_mask.logical_not()

    # Copy-paste images:
    image = image.mul(inverse_paste_alpha_mask).add_(paste_image.mul(paste_alpha_mask))
    # Copy-paste masks:
    masks = masks * inverse_paste_alpha_mask


def proposal(*, image, masks, paste_image, paste_alpha_mask, size1_ne_size2):
    if size1_ne_size2:
        tmp = paste_image.mul_(paste_alpha_mask)
    else:
        tmp = paste_image.mul(paste_alpha_mask)
    inverse_paste_alpha_mask = paste_alpha_mask.logical_not_()

    # Copy-paste images:
    image = tmp.add_(image.mul(inverse_paste_alpha_mask))
    # Copy-paste masks:
    masks = masks.mul(inverse_paste_alpha_mask)


timers = [
    benchmark.Timer(
        stmt="fn(image=image, masks=masks, paste_image=paste_image, paste_alpha_mask=paste_alpha_mask, size1_ne_size2=size1_ne_size2)",
        globals=dict(
            fn=fn,
            image=torch.testing.make_tensor(
                shape,
                dtype=dtype,
                device=device,
                low=0,
                high=1 if dtype.is_floating_point else torch.iinfo(dtype).max,
            ),
            masks=torch.testing.make_tensor(
                (4, *shape),
                dtype=torch.uint8,
                device=device,
                low=0,
                high=2,
            ),
            paste_image=torch.testing.make_tensor(
                shape,
                dtype=dtype,
                device=device,
                low=0,
                high=1 if dtype.is_floating_point else torch.iinfo(dtype).max,
            ),
            paste_alpha_mask=torch.testing.make_tensor(
                shape,
                dtype=torch.bool,
                device=device,
                low=0,
                high=2,
            ),
            size1_ne_size2=size1_ne_size2,
        ),
        sub_label=f"{shape!s:16} / {str(dtype).replace('torch.', ''):7} / {device:4} / size1_ne_size2={size1_ne_size2}",
        description=fn.__name__,
    )
    for fn, shape, dtype, device, size1_ne_size2 in itertools.product(
        [
            current,
            proposal,
        ],
        shapes,
        dtypes,
        devices,
        [False, True],
    )
]

measurements = [timer.blocked_autorange(min_run_time=10) for timer in timers]

comparison = benchmark.Compare(measurements)
comparison.print()

```

</p>
</details>

```
                                                                |  current  |  proposal
1 threads: ----------------------------------------------------------------------------
      (3, 512, 512)    / uint8   / cpu  / size1_ne_size2=False  |   1964.5  |   2155.9 
      (3, 512, 512)    / uint8   / cpu  / size1_ne_size2=True   |   1956.6  |   2169.5 
      (3, 512, 512)    / uint8   / cuda / size1_ne_size2=False  |    142.8  |    140.8 
      (3, 512, 512)    / uint8   / cuda / size1_ne_size2=True   |    143.4  |    139.6 
      (3, 512, 512)    / float32 / cpu  / size1_ne_size2=False  |   2053.3  |   2270.9 
      (3, 512, 512)    / float32 / cpu  / size1_ne_size2=True   |   2052.9  |   2217.7 
      (3, 512, 512)    / float32 / cuda / size1_ne_size2=False  |    196.9  |    197.0 
      (3, 512, 512)    / float32 / cuda / size1_ne_size2=True   |    196.9  |    196.7 
      (5, 3, 512, 512) / uint8   / cpu  / size1_ne_size2=False  |  10575.0  |  10932.7 
      (5, 3, 512, 512) / uint8   / cpu  / size1_ne_size2=True   |  10541.8  |  10887.1 
      (5, 3, 512, 512) / uint8   / cuda / size1_ne_size2=False  |    695.4  |    696.4 
      (5, 3, 512, 512) / uint8   / cuda / size1_ne_size2=True   |    695.2  |    697.6 
      (5, 3, 512, 512) / float32 / cpu  / size1_ne_size2=False  |  14170.3  |  14315.7 
      (5, 3, 512, 512) / float32 / cpu  / size1_ne_size2=True   |  14110.5  |  13735.6 
      (5, 3, 512, 512) / float32 / cuda / size1_ne_size2=False  |    946.1  |    949.2 
      (5, 3, 512, 512) / float32 / cuda / size1_ne_size2=True   |    948.9  |    951.2 

Times are in microseconds (us).
```

- It makes no difference on CUDA
- There is a slowdown on CPU
- Only for the combination `(5, 3, 512, 512) / float32 / cpu  / size1_ne_size2=True` there is a speed-up, of ~400µs. It is likely that the speed-up comes from the `size1_ne_size2=True` condition that I suggested in https://github.com/pytorch/vision/pull/6835#discussion_r1005298828.

I would leave it as is.